### PR TITLE
bpo-27945: Fixed various segfaults with dict.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Based on patches by Duane Griffin and Tim Mitchell.

### CODE REVIEW

File: Lib/test/test_dict.py [hunk 1]
```diff
@@ -1085,6 +1085,87 @@ def test_free_after_iterating(self):
...
+
+            def __eq__(self, other):
+                if len(d) > 1:
+                    d.clear()
+                return False
+
+        d = {}  # this is required to exist so that d can be constructed!
+        d = {X(1): 1, X(2): 2}
+        with self.assertRaises(RuntimeError):
+            dict.fromkeys(d)
```
Comment (encukou) [inline Lib/test/test_dict.py]: These, `test_fromkeys_operator_modifying_dict_operand` and `test_fromkeys_operator_modifying_dict_operand`, are the same, both testing dicts.

Comment (serhiy-storchaka) [reply Lib/test/test_dict.py]: Good catch!

**Review by methane — APPROVED** [review]

File: Objects/dictobject.c [hunk 2]
```diff
@@ -3650,20 +3659,25 @@ dictiter_iternextitem(dictiterobject *di)
...
         result = PyTuple_New(2);
         if (result == NULL)
             return NULL;
+        PyTuple_SET_ITEM(result, 0, key);  /* steals reference */
+        PyTuple_SET_ITEM(result, 1, value);  /* steals reference */
     }
-    Py_INCREF(key);
-    Py_INCREF(value);
-    PyTuple_SET_ITEM(result, 0, key);  /* steals reference */
-    PyTuple_SET_ITEM(result, 1, value);  /* steals reference */
```
Comment (zhangyangyu) [inline Objects/dictobject.c:3666]: @serhiy-storchaka , just my curious, simply moving the two Py_INCREF forward is also enough here right? So any advantage here or just personal style?

Comment (serhiy-storchaka) [reply Objects/dictobject.c:3666]: Hmm... You are right! `Py_INCREF(result)` prevents reentering in that branch or modifying the tuple by other code. Shame on me that I missed such simple solution!

Comment (serhiy-storchaka) [reply Objects/dictobject.c:3666]: But on other hand, `Py_DECREF()` can trigger the garbage collecting. The garbage collector can see the tuple with references to freed objects and expose it to user code via `gc.get_objects()` or `gc.get_referrers()`. Doing some operations with this tuple (`repr()` or `hash()`) can cause a crash. We already encountered with similar issues.

So yes, there are reasons for writing the code in that way.

Comment (zhangyangyu) [reply Objects/dictobject.c:3666]: Ahh, nice point! Stupid me! I didn't recognise it's just like `Py_SETREF(PyTuple_GET_ITEM(result, 0), key)`. But it seems to mean that in your case, users are possible to see a mutable tuple, which smells no good.

Comment (serhiy-storchaka) [reply Objects/dictobject.c:3666]: At that point there is only one owner of the tuple. From user's point of view there is no difference between mutating this tuple or deallocating it and allocating a new tuple at the same memory address.

**Pull request merged**